x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -301,18 +372,56 @@ private void retrieveStackTraces(
         List<List<String>> slicedEventIds = sliced(eventIds, desiredSlices);
         ClusterState clusterState = clusterService.state();
         List<Index> indices = resolver.resolve(clusterState, "profiling-stacktraces", responseBuilder.getStart(), responseBuilder.getEnd());
+
+        // Build a set of unique host IDs.
+        Set<String> uniqueHostIDs = new HashSet<>(responseBuilder.hostEventCounts.size());
+        responseBuilder.hostEventCounts.forEach(hec -> {
+            if (hostsTable.containsKey(hec.hostID) == false) {
+                uniqueHostIDs.add(hec.hostID);
+            }
+        });
```

Comment:
To me this reads as a mix of functional and imperative style. A purely functional style is:

```java
        Set<String> uniqueHostIDs = responseBuilder.hostEventCounts.stream()
            .filter((h -> hostsTable.containsKey(h.hostID) == false))
            .map(h -> h.hostID)
            .collect(Collectors.toSet());
```

Alternatively (if you prefer an imperative approach) I propose we use a regular for loop to iterate over `responseBuilder.hostEventCounts`.

Also, how about we rename `uniqueHostIDs` to `unknownHostIDs`, `newHostIDs` or something like that as it actually contains host ids that we did not cache yet?